### DISCUSSION TITLE
Is 'compile failed' a cache miss?

### DISCUSSION THREAD

**Discussion by AlexanderLanin** [ccache/ccache#892] [Questions & answers] [answered]
Exemplary ccache stats:

```
cache hit (direct)                   291
cache hit (preprocessed)               1
cache miss                             0
cache hit rate                    100.00 %
compile failed                       103
preprocessor error                    51
```

Shouldn't those failures be counted as `cache miss (compile failed)` and `cache miss (preprocessor error)`?
With a hit rate of 65,47%.

Is that a bug or a feature or is my understanding wrong?
Feels like a bug in the way statistics are reported and hit rate is calculated.

My understanding is that `compile failed` and `preprocessor error` are not _technically_ cache misses since they don't even participate in caching.
On the other hand they are looked up, there is cache miss, and then there is the error.

Anyway suggesting a 100% cache hit rate seems misleading. After all the preprocessor/compiler has been called 154 times. Which is not what someone would expect when they see 100% cache hit rate.

**Comment by jrosdahl** [comment, marked as answer]
For reference, the current way of calculating hit rate was introduced by #94/#109. I find it hard to call it a bug, but I agree that it's possible to argue that hit rate could be defined in different ways. As you say, some may find it intuitive to include preprocessor and compilation failures in the hit rate, and some may find it intuitive to include "compiler produced stdout", "could not write to output file", "compiler produced no output", "compiler produced empty output", etc.

> Which is not what someone would expect when they see 100% cache hit rate.

I don't think I'm alone in finding the current hit rate calculation appropriate, or at least not inappropriate. Here's my intuition: After clearing statistics and recompiling, 100% hit rate means "I have done everything I can to increase hit rate". If compilation failures due to build system feature probing were to be included, the hit rate would be lower and it would appear as if there are still some things left to do to increase hit rate.

**Comment by AlexanderLanin** [comment]
I do like that definition:

> After clearing statistics and recompiling, 100% hit rate means "I have done everything I can to increase hit rate".
